File: src/tui/progressStore.ts
export interface ProgressData {
  progress: string;
  percent: number;
  speed: string;
  eta: string;
  lastVideoSeq?: number;
  lastAudioSeq?: number;
  totalVideoSeq?: number;
  totalAudioSeq?: number;
  totalChatMessages?: number;
  chatStatus?: string;
}

// Module-level — no React involvement, zero allocations on read
const store = new Map<string, ProgressData>();

export function setProgress(jobId: string, data: ProgressData): void {
  store.set(jobId, data);
}

export function getProgress(jobId: string): ProgressData | undefined {
  return store.get(jobId);
}

export function deleteProgress(jobId: string): void {
  store.delete(jobId);
}

export function clearProgress(): void {
  store.clear();
}

File: internal/tui/progress_store.go
package tui

// ProgressData stores high-frequency download progress without triggering
// TUI re-sorts. Only status changes cause re-renders; progress reads are
// zero-cost lookups from this store during View().
type ProgressData struct {
	Progress          string
	Percent           float64
	Speed             string
	ETA               string
	LastVideoSeq      *int
	LastAudioSeq      *int
	TotalVideoSeq     *int
	TotalAudioSeq     *int
	TotalChatMessages *int
	ChatStatus        string
}

// ProgressStore is a map of job ID -> progress data.
// All access occurs on the main BubbleTea goroutine (Update/View),
// so no synchronization is needed.
type ProgressStore struct {
	items map[string]*ProgressData
}

// NewProgressStore creates a new progress store.
func NewProgressStore() *ProgressStore {
	return &ProgressStore{
		items: make(map[string]*ProgressData),
	}
}

// Set updates progress data for a job.
func (s *ProgressStore) Set(jobID string, data *ProgressData) {
	s.items[jobID] = data
}

// Get returns progress data for a job, or nil.
func (s *ProgressStore) Get(jobID string) *ProgressData {
	return s.items[jobID]
}

// Delete removes progress data for a job.
func (s *ProgressStore) Delete(jobID string) {
	delete(s.items, jobID)
}

// Clear removes all entries.
func (s *ProgressStore) Clear() {
	s.items = make(map[string]*ProgressData)
}
